db/src/types.rs

```diff
@@ -0,0 +1,251 @@
+    /// lookup-refs and will automatically upsert where appropriate.
+    pub unique_identity: bool,
+    /// `true` if this attribute is automatically indexed, i.e., it is `:db/indexing true`.
+    pub index: bool,
+    /// `true` if this attribute is automatically fulltext indexed, i.e., it is `:db/fulltext true`.
+    ///
+    /// Fulltext attributes always have string values.
+    pub fulltext: bool,
+    /// `true` if this attribute is a component, i.e., it is `:db/isComponent true`.
+    ///
+    /// Component attributes always have value type `Ref`.
+    ///
+    /// They are used to compose entities from component sub-entities: they are fetched recursively
+    /// by pull expressions, and they are automatically recursively deleted where appropriate.
+    pub component: bool,
+}
+
+impl Default for Attribute {
+    fn default() -> Attribute {
+        Attribute {
+            // There's no particular reason to favour one value type, so Ref it is.
+            value_type: ValueType::Ref,
+            fulltext: false,
+            index: false,
+            multival: false,
+            unique_value: false,
+            unique_identity: false,
+            component: false,
+        }
+    }
+}
+
+/// Map `String` idents (`:db/ident`) to positive integer entids (`1`).
+pub type IdentMap = BTreeMap<String, Entid>;
+
+/// Map positive integer entids (`1`) to `String` idents (`:db/ident`).
+pub type EntidMap = BTreeMap<Entid, String>;
+
+/// Map attribute entids to `Attribute` instances.
+pub type SchemaMap = BTreeMap<i64, Attribute>;
+
+/// Represents a Mentat schema.
+///
+/// Maintains the mapping between string idents and positive integer entids; and exposes the schema
+/// flags associated to a given entid (equivalently, ident).
+///
+/// TODO: consider a single bi-directional map instead of separate ident->entid and entid->ident
+/// maps.
+#[derive(Clone,Debug,Default,Eq,Hash,Ord,PartialOrd,PartialEq)]
+pub struct Schema {
+    /// Map entid->ident.
+    ///
+    /// Invariant: is the inverse map of `ident_map`.
+    pub entid_map: EntidMap,
+    /// Map ident->entid.
+    ///
+    /// Invariant: is the inverse map of `entid_map`.
+    pub ident_map: IdentMap,
+    /// Map entid->attribute flags.
+    ///
+    /// Invariant: key-set is the same as the key-set of `entid_map` (equivalently, the value-set of
+    /// `ident_map`).
+    pub schema_map: SchemaMap,
+}
+
+impl Schema {
+    pub fn new(ident_map: IdentMap, schema_map: SchemaMap) -> Schema {
+        let entid_map: EntidMap = ident_map.iter().map(|(k, v)| (v.clone(), k.clone())).collect();
+        Schema {
+            ident_map: ident_map,
+            entid_map: entid_map,
+            schema_map: schema_map,
+        }
+    }
+}
+
+/// Represents the metadata required to query from, or apply transactions to, a Mentat store.
+#[derive(Clone,Debug,Default,Eq,Hash,Ord,PartialOrd,PartialEq)]
```

Comment:
Perhaps leave a pointer to <https://github.com/mozilla/mentat/wiki/Thoughts:-modeling-db-conn-in-Rust> here, because eventually people will wonder why `DB` looks the way it does…